lib/datastore/datastore_container.es6.js

```diff
@@ -0,0 +1,205 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+require('../confluence/api_velocity_data.es6.js');
+require('../confluence/browser_metric_data.es6.js');
+require('../dao_container.es6.js');
+require('../web_apis/release.es6.js');
+require('../web_apis/release_interface_relationship.es6.js');
+require('../web_apis/web_interface.es6.js');
+
+foam.CLASS({
+  name: 'DatastoreContainer',
+  package: 'org.chromium.apis.web',
+  extends: 'org.chromium.apis.web.DAOContainer',
+
+  documentation: `Controller acting as container for datastore context. Note
+      that this is component is Node JS-only because it requires
+      com.google.net.node.Google2LOAuthAgent.`,
+
+  requires: [
+    'com.google.cloud.datastore.BatchMutationDatastoreDAO',
+    'com.google.net.node.Google2LOAuthAgent',
+    'foam.nanos.log.ConsoleLogger',
+    'org.chromium.apis.web.ApiVelocityData',
+    'org.chromium.apis.web.BrowserMetricData',
+    'org.chromium.apis.web.Release',
+    'org.chromium.apis.web.ReleaseWebInterfaceJunction',
+    'org.chromium.apis.web.WebInterface',
+  ],
+  exports: ['gcloudProjectId'],
+
+  classes: [
+    {
+      name: 'HTTPRequest',
+      extends: 'foam.net.BaseHTTPRequest',
+
+      documentation: `Wrapper class for defining HTTPRequest delegate chain
+          under this controller. The desired chain is as follows:
+
+          RetryHTTPRequest
+              -> (agent-contextualized) AuthHTTPRequest
+              -> (default) HTTPRequest.
+
+          When no authAgent is specified, the AuthHTTPRequest is left out.
+
+          This ensures that requests that fail at the service level will be
+          retried a reasonable number of times, and that all requests requiring
+          credentials are appropriately authorized.`,
+
+      requires: [
+        'foam.net.BaseHTTPRequest',
+        'foam.net.RetryHTTPRequest',
+      ],
+      imports: [ 'authAgent?' ],
+
+      properties: [
+        {
+          class: 'Proxy',
+          of: 'foam.net.BaseHTTPRequest',
+          name: 'delegate',
+          factory: function() {
+            var BaseHTTPRequest = this.BaseHTTPRequest;
+            var RetryHTTPRequest = this.RetryHTTPRequest;
+
+            var baseProps = {};
+            this.BaseHTTPRequest.getAxiomsByClass(foam.core.Property)
+                .forEach(prop => baseProps[prop.name] = this[prop.name]);
+
+            // No auth agent => retry decorator only.
+            if (!this.authAgent) {
+              return RetryHTTPRequest.create(Object.assign({}, baseProps, {
+                delegate: BaseHTTPRequest.create(baseProps, this),
+              }), this);
+            }
+
+            var AuthHTTPRequest =
+                  this.authAgent.__subContext__.lookup('foam.net.HTTPRequest');
+            return RetryHTTPRequest.create(Object.assign({}, baseProps, {
```

Comment:
I don't know if this is better, but you could do something like this to avoid having two object construction paths:
```
var next = BaseHTTPRequest
if (authAgent) {
  next = AuthHTTP(delegate: next)
}
next = RetryHTTP(delegate: next)
```